File: army/JavaArmy/src/com/maksklaban/armymax/units/Unit.java
package com.maksklaban.armymax.units;

import com.maksklaban.armymax.exceptions.*;
import com.maksklaban.armymax.state.*;
import com.maksklaban.armymax.attack.*;
import java.util.Set;
import java.util.HashSet;
import java.util.Iterator;
import java.lang.ref.*;

public class Unit extends Ability implements Subject, Observer {
    protected Set<WeakReference<Unit>> observerList;
    protected ReferenceQueue<Unit> queue;
    protected State state;
    protected State normalState;
    protected WolfState wolfState;
    protected NormalAttack physicalAttack;
    protected NormalAttack temp;
    protected String nickName;
    protected boolean isUndead;

    public Unit(String nickName, int damage, int hitPoints, String title, boolean isUndead, boolean isVampire, boolean isWerewolf) throws CloneNotSupportedException {
        super(isVampire, isWerewolf);
        this.queue = new ReferenceQueue<Unit>();
        this.observerList = new HashSet<WeakReference<Unit>>();
        this.nickName = nickName;
        this.state = new State(hitPoints, damage, title);
        this.normalState = this.state.clone();
        this.wolfState = new WolfState(hitPoints, damage, title);
        this.physicalAttack = new DefaultAttack();
        this.isUndead = isUndead;
    }

    public Unit(String nickName, int damage, int hitPoints, String title, boolean isUndead) throws CloneNotSupportedException {
        this(nickName, damage, hitPoints, title, isUndead, false, false);
    }

    public Unit(String nickName, int damage, int hitPoints, String title) throws CloneNotSupportedException {
        this(nickName, damage, hitPoints, title, false, false, false);
    }

    public String getNickName() {
        return this.nickName;
    }

    public int getDamage() {
        return this.state.getDamage();
    }

    public int getHitPoints() {
        return this.state.getHitPoints();
    }

    public int getHitPointsLimit() {
        return this.state.getHitPointsLimit();
    }

    public boolean getIsUndead() {
        return this.isUndead;
    }

    public String getTitle() {
        return this.state.getTitle();
    }

    public State getState() {
        return this.state;
    }

    public void update(Unit other) throws ArmyException {}

    public void attach(Unit another) {
        WeakReference<Unit> buff = new WeakReference<Unit>(another, queue);

        this.observerList.add(buff);
    }


    public void notifyObserver() throws ArmyException {
        Iterator it = this.observerList.iterator();

        for ( ; it.hasNext(); ) {
            WeakReference temp = (WeakReference)it.next();
            Unit buff = (Unit)temp.get();

            buff.update(this);
        }
    }

    public void takeDamage(int dmg) throws ArmyException {
        this.state.takeDamage(dmg);

        if ( this.getHitPoints() == 0 ) {
            this.notifyObserver();
        }
    }

    public void takeMagDamage(int dmg) throws ArmyException {
        this.state.takeMagDamage(dmg);
    }

    public void takePhysicalDamage(int dmg) throws ArmyException {
        this.takeDamage(dmg);
    }

    public void addHitPoints(int hp) throws ArmyException {
        this.state.addHitPoints(hp);
    }

    public void transformInToWolf() throws ArmyException {
        int temp = this.getHitPoints() + (this.getHitPoints() / 2);

        if ( this.getIsWerewolf() && this.getHitPoints() > 0 ) {
            this.state = this.wolfState;
            this.state.setCurrentHp(temp);
        } else {
            throw new UnitCantTransformException();
        }
    }

    public void transformBack() throws ArmyException {
        int temp = this.getHitPoints() - (this.getHitPoints() / 2);

        if ( this.getTitle() == "WolfState" && this.getHitPoints() > 0 ) {
            this.state = this.normalState;
            this.state.setCurrentHp(temp);
        } else {
            throw new UnitCantTransformException();
        }
    }

    public void getVampireAttack() throws ArmyException {
        if ( this.getIsVampire() ) {
            this.temp = this.physicalAttack;
            this.physicalAttack = new AttackVampire();
        } else {
            throw new AbilityIsNotAvaibleException();
        }
    }

    public void getNormalAttack() throws ArmyException {
        if ( this.getIsVampire() ) {
            this.physicalAttack = this.temp;
        } else {
            throw new AbilityIsNotAvaibleException();
        }
    }

    public void attack(Unit attacker, Unit enemy) throws ArmyException {
        this.physicalAttack.attack(this, enemy);
    }

    public void counterAttack(Unit attacker, Unit enemy) throws ArmyException {
        this.physicalAttack.counterAttack(this, enemy);
    }

    public String toString() {
        StringBuilder out = new StringBuilder(this.getNickName());
        out.append(" " + super.toString() + this.getState());

        if ( this.getIsUndead() ) {
            out.append(", Undead");
        }

        return out.toString();
    }
}

File: army/Unit/Unit.cpp
#include "Unit.h"

Unit::Unit(std::string nickName, int damage, int hitPoints, std::string title, bool isUndead, bool isVampire, bool isWerewolf) : nickName(nickName), isUndead(isUndead), physicalAttack(new DefaultAttack() ), state(new State(hitPoints, damage, title)), Ability(isVampire, isWerewolf), Subject(), Observer() {
    this->normalState = new State(hitPoints, damage, title);
    this->wolfState = new WolfState(hitPoints, damage, title);
}
 
Unit::~Unit() {
    delete this->normalState;
    delete this->wolfState;
    delete this->state;
    delete this->physicalAttack;
    delete this->temp;
}

const std::string& Unit::getNickName() const {
    return this->nickName;
}

const int Unit::getDamage() const {
    return this->state->getDamage();
}

const int Unit::getHitPoints() const {
    return this->state->getHitPoints();
}

const int Unit::getHitPointsLimit() const {
    return this->state->getHitPointsLimit();
}

const bool Unit::getIsUndead() const {
    return this->isUndead;
}

const std::string& Unit::getTitle() const {
    return this->state->getTitle();
}

const State& Unit::getState() const {
    return *(this->state);
}

void Unit::takeDamage(int dmg) {
    this->state->takeDamage(dmg);

    if ( this->getHitPoints() == 0 ) {
        this->notify();
    }
}

void Unit::takeMagDamage(int dmg) {
    this->state->takeMagDamage(dmg);
}

void Unit::takePhysicalDamage(int dmg) {
    this->takeDamage(dmg);
}

void Unit::addHitPoints(int hp) {
    this->state->addHitPoints(hp);
}

void Unit::turnInVampire(Unit* enemy) {
    Ability::turnInVampire(enemy);

    enemy->isUndead = true;
}

void Unit::transformInToWolf() {
    int temp = this->getHitPoints() * 1.5;

    if ( this->isWerewolf && this->getHitPoints() > 0 ) {
        this->state = this->wolfState;
        this->state->setCurrentHp(temp);
    } else {
        throw UnitCantTransformException();
    }
}

void Unit::transformBack() {
    int temp = this->getHitPoints() / 1.5;

    if ( this->getTitle() == "WolfForm" && this->getHitPoints() > 0 ) {
        this->state = this->normalState;
        this->state->setCurrentHp(temp);
    } else {
        throw UnitCantTransformException();
    }
}

void Unit::getVampireAttack() {
    if ( this->getIsVampire() ) {
        this->temp = this->physicalAttack;
        this->physicalAttack = new AttackVampire();
    } else {
        throw AbilityIsNotAvaibleException();
    }
}

void Unit::getNormalAttack() {
    if ( this->getIsVampire()) {
        delete this->physicalAttack;
        this->physicalAttack = this->temp;
    } else {
        throw AbilityIsNotAvaibleException();
    }
}

void Unit::attack(Unit* attacker, Unit* enemy) {
    this->physicalAttack->attack(this, enemy);
}

void Unit::counterAttack(Unit* attacker, Unit* enemy) {
    this->physicalAttack->counterAttack(this, enemy);
}

std::ostream& operator<<(std::ostream& out, const Unit& unit) {
    out << unit.getNickName() << " " << (Ability)unit << unit.getState();
    if ( unit.getIsUndead() ) {
        out << ", Undead";
    }
    return out;
}
